### PR TITLE
nvidia_temp: add format, use helpers

### PR DESCRIPTION
**Pull request by lasers**
EDIT: Closes https://github.com/ultrabug/py3status/issues/699. 
EDIT: Closes https://github.com/ultrabug/py3status/issues/754.

This adds format: string to print (default 'GPU: {format_temp}')
This adds format_temp: string to print (default '{temp}°C')
This adds ~~format~~ string_unavailable: string to print if unavailable (default 'N/A')

I didn't test Unavailable (How?)
I didn't test format_separator (Extra GPU required).

Will break. Hold this for v3.4 or what?

Ongoing efforts to `FORMAT ALL THE THINGS!` (meme). https://github.com/ultrabug/py3status/issues/98

### CODE REVIEW

**Comment by tobes** [comment]
We cannot just change the options it breaks thing for users.  Things can be done using the deprecation functionality but that is not always simple.

Breaking things for users is never good!

**Comment by lasers** [comment]
Got it. What to do here? Just filling in some of the missing formats. https://github.com/ultrabug/py3status/issues/98

This is slightly tricky. Can't just do format = "{prefix}{temp}{unit}" because of separator due to possible extra GPUs resulting something like `GPU: 60 60°C` for some users.

**Comment by tobes** [comment]
> This is slightly tricky. 

Yep sometimes it is.  We do something slightly similar in say module `clock` which uses 2 formats to do this.

**Comment by lasers** [comment]
@tobes No way around this. Also, I think it'll be easier to recall if I switch out `msg_temp` with `gpu_temp`. It looks like sysdata use `{cpu_temp}` already so this should be okay. Anything else you want to do here?

File: py3status/modules/nvidia_temp.py [hunk 1]
```diff
@@ -43,33 +59,21 @@ def nvidia_temp(self):
...
-                )
-                data.append(fmt_str)
-
-            output = "{format_prefix}{data}".format(
-                format_prefix=self.format_prefix,
-                data=self.temp_separator.join(data)
-            )
+                data.append(self.py3.safe_format(self.gpu_temp, {'temp': temp}))
 
+            output = "{data}".format(data=self.icon_separator.join(data))
```
Comment (tobes) [inline py3status/modules/nvidia_temp.py]: here we want to build a composite to pass as `{gpu_temps}` which is basically your data but with `{'full_text': self.icon_separator}` between each item.

probably easiest to add during your loop and then remove the last item.

maybe in future the separator should be a format so we can colour it? for now I'd just keep it plain text.

File: py3status/modules/nvidia_temp.py [hunk 2]
```diff
@@ -1,17 +1,21 @@
...
 Configuration parameters:
     cache_timeout: how often we refresh this module in seconds (default 10)
-    format_prefix: a prefix for the output. (default 'GPU: ')
-    format_units: the temperature units. Will appear at the end. (default '°C')
-    temp_separator: the separator char between temperatures (only if more than
-        one GPU) (default '|')
+    format: display format for nvidia_temp (default 'GPU: {gpu_temp}')
+    gpu_temp: display format for temperature(s) (default '{temp}°C')
+    icon_separator: display separator char (only if more than one) (default '|')
+    msg_unavailable: display when unavailable (default 'GPU: N/A')
```
Comment (tobes) [inline py3status/modules/nvidia_temp.py]: this should wait for #660

File: py3status/modules/nvidia_temp.py [hunk 3]
```diff
@@ -32,9 +36,21 @@ class Py3status:
     """
     # configuration parameters
     cache_timeout = 10
-    format_prefix = "GPU: "
-    format_units = u"°C"
-    temp_separator = '|'
+    format = "GPU: {gpu_temp}"
+    gpu_temp = u"{temp}°C"
```
Comment (tobes) [inline py3status/modules/nvidia_temp.py]: I'd call this `format_temp` as it uses formatter and has placeholder. but default `'{temp}{temp_unit}'` and have the °C set by `temp_unit` like we are doing in #555 

for format I'd have the placeholder `{gpu_temps}` as there may be multiple ones.

we can create the output for `gpu_temps`

Comment (tobes) [inline py3status/modules/nvidia_temp.py]: actually looks like you can't choose temperature scale so maybe just hardcode as you are

Comment (lasers) [inline py3status/modules/nvidia_temp.py]: Heh. I was changing things. Final say? Keep everything as-is?

Comment (tobes) [inline py3status/modules/nvidia_temp.py]: no just don't have the temp_unit

Comment (tobes) [inline py3status/modules/nvidia_temp.py]: unless you want to do the conversions but again I do that as a different PR. who uses °F anyway, oh yeah...

Comment (lasers) [inline py3status/modules/nvidia_temp.py]: I feel `format_temp` says (format temporary) message.
I use `format_gpu_temp` to explicitly say (format gpu temp).

Comment (tobes) [inline py3status/modules/nvidia_temp.py]: sounds good

Comment (lasers) [inline py3status/modules/nvidia_temp.py]: Finished. Put a tag up after a quick lookie. Will look at other modules.

File: py3status/modules/nvidia_temp.py [hunk 4]
```diff
@@ -43,33 +59,21 @@ def nvidia_temp(self):
...
-                )
-                data.append(fmt_str)
-
-            output = "{format_prefix}{data}".format(
-                format_prefix=self.format_prefix,
-                data=self.temp_separator.join(data)
-            )
+                data.append(self.py3.safe_format(self.format_gpu_temp, {'temp': temp}))
 
+            output = "{data}".format(data=self.icon_separator.join(data))
```
Comment (tobes) [inline py3status/modules/nvidia_temp.py]: this bit still needs fixing like I described.

I've a friend coming over now so talk more another day

Comment (lasers) [inline py3status/modules/nvidia_temp.py]: What needs to be done here? I would like to get this out of way asap.

Comment (tobes) [inline py3status/modules/nvidia_temp.py]: something like this
```
for temp in temps:
    data.append(self.py3.safe_format(self.format_temp, {'temp': temp}))
    data.append({'full_text': self.icon_separator})

if data:
    data = data[:-1]
```

Comment (lasers) [inline py3status/modules/nvidia_temp.py]: `returned "NV: ['61°C']"`

File: py3status/modules/nvidia_temp.py [hunk 5]
```diff
@@ -43,33 +59,25 @@ def nvidia_temp(self):
...
 
-            output = "{format_prefix}{data}".format(
-                format_prefix=self.format_prefix,
-                data=self.temp_separator.join(data)
-            )
+            if data:
+                data = data[:-1]
 
+            full_text = self.py3.safe_format(self.format,
+                                             {'format_temp': data, 'temp': data})
```
Comment (tobes) [inline py3status/modules/nvidia_temp.py]: `'temp'` should not be here, it is for format_temp only

File: py3status/modules/nvidia_temp.py [hunk 6]
```diff
@@ -1,17 +1,21 @@
...
-    temp_separator: the separator char between temperatures (only if more than
-        one GPU) (default '|')
+    format: display format for nvidia_temp (default 'GPU: {format_temp}')
+    format_temp: display format for gpu temperature(s) (default '{temp}°C')
+    icon_separator: display separator (only if more than one) (default '|')
+    msg_unavailable: display when unavailable (default 'GPU: N/A')
+
+Format placeholders:
+    {format_temp} display current gpu temperature(s)
+    {temp} display format for gpu temperature
```
Comment (tobes) [inline py3status/modules/nvidia_temp.py]: `{temp}` should be in its own section as an example see https://github.com/ultrabug/py3status/blob/master/py3status/modules/diskdata.py

File: py3status/modules/nvidia_temp.py (hunk 4 above)
Comment (tobes) [inline py3status/modules/nvidia_temp.py]: Actually this should work better
```
for temp in temps:
    data.append(self.py3.safe_format(self.format_temp, {'temp': temp}))

data = self.py3.composite_join(self.icon_separator, data)
```

`temp_data` may be a better name than `data`

Sorry I can't test this module

File: py3status/modules/nvidia_temp.py [hunk 7]
```diff
@@ -32,9 +38,21 @@ class Py3status:
...
+    class Meta:
+        deprecated = {
+            'rename': [
+                {
+                    'param': 'temp_separator',
+                    'new': 'icon_separator',
+                    'msg': 'obsolete parameter use `icon_separator`',
+                },
+            ],
+        }
```
Comment (tobes) [inline py3status/modules/nvidia_temp.py]: `format_units` needs to be dealt with.  It's a bit of a pain but basically we will use 'function' to create a new `format_temp` (unless the user supplied one) and then remove `format_units`

File: py3status/modules/nvidia_temp.py [hunk 8]
```diff
@@ -43,33 +61,24 @@ def nvidia_temp(self):
         if temps != []:
             data = []
             for temp in temps:
-                fmt_str = "{temp}{format_units}".format(
-                    temp=temp,
-                    format_units=self.format_units
-                )
-                data.append(fmt_str)
+                data.append(self.py3.safe_format(self.format_temp, {'temp': temp}))
+                data.append({'full_text': self.icon_separator})
```
Comment (tobes) [inline py3status/modules/nvidia_temp.py]: this still needs fixing (my mistake)
```
for temp in temps:
    data.append(self.py3.safe_format(self.format_temp, {'temp': temp}))

data = self.py3.composite_join(self.icon_separator, data)
```

also means we can lose 

```
if data:
    data = data[:-1]
```

File: py3status/modules/nvidia_temp.py [hunk 9]
```diff
@@ -1,17 +1,23 @@
...
 Configuration parameters:
     cache_timeout: how often we refresh this module in seconds (default 10)
-    format_prefix: a prefix for the output. (default 'GPU: ')
-    format_units: the temperature units. Will appear at the end. (default '°C')
-    temp_separator: the separator char between temperatures (only if more than
-        one GPU) (default '|')
+    format: display format for nvidia_temp (default 'GPU: {format_temp}')
+    format_temp: display format for gpu temperature(s) (default '{temp}°C')
+    icon_separator: display separator (only if more than one) (default '|')
+    msg_unavailable: display when unavailable (default 'GPU: N/A')
```
Comment (tobes) [inline py3status/modules/nvidia_temp.py]: @lasers You know what I'm going to say here, but also, this changes the behaviour. the unavailable message has a default of `format_prefix + 'N/A'` which needs fixing anyhow.

Till we have consensus on errors #660 and more.  I'd also not make it available for the user to change

Comment (lasers) [inline py3status/modules/nvidia_temp.py]: Github should just stop collapse all outdated comments. Display everything as-is. I removed the prefix, but will leave `N/A`. I think people won't use it (the module) at all if they found out that it can't read the temperature. Just return direct `N/A` instead?

**Comment by tobes** [comment]
This patch will fix up the formats for users who have bits set.

```
diff --git a/py3status/modules/nvidia_temp.py b/py3status/modules/nvidia_temp.py
index 39d1f19..7ccf408 100644
--- a/py3status/modules/nvidia_temp.py
+++ b/py3status/modules/nvidia_temp.py
@@ -40,6 +40,23 @@ class Py3status:
     format_temp = u"{temp}°C"
     temp_separator = '|'
 
+    class Meta:
+
+        def deprecate_function(config):
+            out = {}
+            if 'format_units' in config:
+                out['format_temp'] = u'{{temp}}{}'.format(config['format_units'])
+            if 'format_prefix' in config:
+                out['format'] = u'{}{{format_temp}}'.format(config['format_prefix'])
+            return out
+
+
+        deprecated = {
+            'function': [
+                {'function': deprecate_function},
+            ],
+        }
+
     def nvidia_temp(self):
         if not self.py3.check_commands('nvidia-smi'):
             return {
```
